### ISSUE TITLE
gradient of segment_sum gives nan 

### ISSUE THREAD

**Issue by thisiscam** [google/jax#14179]
### Description

I noticed that `segment_sum` seems to have [a suspiciously similar problem to `jnp.where`](https://jax.readthedocs.io/en/latest/faq.html?highlight=nan#gradients-contain-nan-where-using-where).

Here is the code that produces the issue:
```python
import jax
import jax.ops
import jax.numpy as jnp

x = jnp.array([1., 2., 3.])
a = jnp.array([1., 1., jnp.nan])
segment_ids = jnp.array([0, 1, -1], dtype=jnp.int32)

def f(x):
  return jax.ops.segment_sum(a * x, segment_ids, num_segments=2)[0]

print(jax.grad(f)(x)) # prints [ 1.  0. nan], while could have been [ 1.  0. 0.]
```

The issue is that when an entry of `segment_id` is `-1`, the corresponding array entry should not contribute to the gradient; however, the current behavior "leaks" the `nan` into the gradient. 

Is the current behavior intended? 
If so, I wasn't able to find information about this---perhaps this could go into the FAQ section of the docs, highlighted similar to `jnp.where`? 

### What jax/jaxlib version are you using?

jax v0.4.1

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.9, macOS

**Comment by jakevdp** [comment]
Symbolically, if you have something that looks like:

$f(a, x) = g(a \cdot x)$

then by the chain rule, the gradient with respect to x is

$\frac{d}{dx}f(a, x) = a * \frac{d}{dx} g(a\cdot x)$

Here, you've set up the problem so that one element of `a` is `NaN`. Because `NaN` multiplied by anything returns `NaN`, the result will include `NaN`. Does that make sense?

**Comment by thisiscam** [comment]
I see, that makes total sense! And I recognize that this is a different problem from the `jnp.where` issue. 
Thank you!
